feat: Verify email of people setting a meeting with you

rkreddy99 opened the pull request:
## What does this PR do?

Allows event owner to make booker verify their email before scheduling event.

Fixes #9901 

 Loom Video: https://www.loom.com/share/3e9f816ffb6148b09de906080c9fe2fb

## Requirement/Documentation

<!-- Please provide all documents that are important to understand the reason of that PR. -->

- Added "requiresBookerEmailVerification" in database.

## Type of change

<!-- Please delete bullets that are not relevant. -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Toggle on the Requires Email Verification in advanced settings for an event.
- Now when the booker tries to book the even they are being sent an code to verify.

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

Commit:
booker email verification changes

Commit:
name type fix

Udit-takkar commented on packages/features/bookings/Booker/components/BookEventForm/EmailVerificationModal.tsx:
```diff
@@ -0,0 +1,80 @@
+  useEffect(() => clearErrors("emailVerificationCode"), [])
+
+  return (
+    <Dialog open={visible} onOpenChange={onCancel}>
+      <DialogContent
+        type={undefined}
+        enableOverflow
+        className="[&_.modalsticky]:border-t-subtle [&_.modalsticky]:bg-default max-h-[80vh] pb-0 [&_.modalsticky]:sticky [&_.modalsticky]:bottom-0 [&_.modalsticky]:left-0 [&_.modalsticky]:right-0 [&_.modalsticky]:-mx-8 [&_.modalsticky]:border-t [&_.modalsticky]:px-8 [&_.modalsticky]:py-4">
+        <h1 className="font-cal text-emphasis text-xl leading-5">{t("verify_email_email_header")}</h1>
+        
+        <FormProvider {...methods}>    
+          <form onSubmit={methods.handleSubmit(onSubmit)} data-testid="booker-email-verification-form" className="flex flex-col items-end py-5">
+            <TextField  
+              id="verification-code"
+              label={t("email_verification_code")}
+              placeholder={t("email_verification_code_placeholder")}
+              autoComplete="off"
+              required={true}
+              className="mb-5"
+              containerClassName="w-full"
+              error={formState.errors.emailVerificationCode?.message}
+              {...register("emailVerificationCode")}
+            />
+            <Button
+             type="submit"
+             color="primary"
+             className="max-w-fit"
+             disabled={formState.isSubmitting}
+            >Verify
+            </Button>
+          </form>
+        </FormProvider>
+      </DialogContent>
+    </Dialog>
```
We already have a verify email dialog. It's in CreateANewOrganizationForm with the name VerifyCodeDialog. maybe you can create a reusable component and use it in both of the places

<img width="1440" alt="Screenshot 2023-07-23 at 4 30 19 PM" src="https://github.com/calcom/cal.com/assets/53316345/25329453-6fea-4b2a-bd4e-84be59bd1ccf">

Udit-takkar commented on packages/features/auth/lib/verifyEmail.ts:
```diff
@@ -54,3 +54,37 @@ export const sendEmailVerification = async ({ email, language, username }: Verif
 };
+
+export const sendEmailVerificationByCode = async ({ email, language, username }: VerifyEmailType) => {
+  const token = randomBytes(3).toString("hex");
```
we use totp library to generate OTP code. check in create.handler file in packages/trpc/server/routers/viewer/organizations/create.handler.ts

rkreddy99 replied:
will check this out.

Udit-takkar commented on packages/trpc/server/routers/viewer/auth/verifyToken.handler.ts:
```diff
@@ -0,0 +1,37 @@
+  input: TVerifyTokenSchema;
+};
+
+export const verifyTokenHandler = async ({ input }: VerifyTokenOptions) => {
```
this looks similar to packages/trpc/server/routers/viewer/organizations/verifyCode.handler.ts. we can maybe use verifycode.handler instead of creating new one and verificationToken table is used to store inviteTokens

Udit-takkar replied:
@rkreddy99  we don't need this function. we can use verifyCode.handler

Udit-takkar reviewed (changes requested):
@rkreddy99 Great work. Left few comments in the code.

Instead of verify email as button text. we should show "continue".

<img width="1039" alt="Screenshot 2023-07-23 at 4 45 27 PM" src="https://github.com/calcom/cal.com/assets/53316345/d88a20b8-e845-4b27-ab4b-0f08bddfb376">

and after code has been entered we should also trigger the confirm button so user doesn't have to click again on the button.

<img width="1039" alt="Screenshot 2023-07-23 at 4 47 43 PM" src="https://github.com/calcom/cal.com/assets/53316345/df7686ed-e7a8-4290-ada2-17edbc0c0f65">

rkreddy99 commented:
> @rkreddy99 Great work. Left few comments in the code.
> 
> Instead of verify email as button text. we should show "continue".
> 
> <img alt="Screenshot 2023-07-23 at 4 45 27 PM" width="1039" src="https://user-images.githubusercontent.com/53316345/255389604-d88a20b8-e845-4b27-ab4b-0f08bddfb376.png">
> and after code has been entered we should also trigger the confirm button so user doesn't have to click again on the button.
> 
> <img alt="Screenshot 2023-07-23 at 4 47 43 PM" width="1039" src="https://user-images.githubusercontent.com/53316345/255389716-df7686ed-e7a8-4290-ada2-17edbc0c0f65.png">

I considered doing that but lets say the user wants to change something after the verification, and that's the reason I stored the verifiedEmail in the store as well so that once the user verifies their email, they can still go back and change something without the need for verification again.

Udit-takkar commented:
> I considered doing that but lets say the user wants to change something after the verification, and that's the reason I stored the verifiedEmail in the store as well so that once the user verifies their email, they can still go back and change something without the need for verification again.

Ah okay. we can two button in modal "cancel" and "Verify Email". Cancel button would just close the modal and not submit the booking form but verify email would submit the form if code is verified

PeerRich commented:
this is super cool @rkreddy99, super useful and great UI implementation.

I wonder if we should use our "PIN" component that we use for 2FA to make the UI a bit nicer: 
![image](https://github.com/calcom/cal.com/assets/8019099/b8013c62-4964-4b13-8612-06af56773016)

sean-brydon commented:
> this is super cool @rkreddy99, super useful and great UI implementation.
> 
> I wonder if we should use our "PIN" component that we use for 2FA to make the UI a bit nicer: ![image](https://user-images.githubusercontent.com/8019099/255556407-b8013c62-4964-4b13-8612-06af56773016.png)

We do already have the pin part of this created as a component IIRC

Commit:
use totp and code

rkreddy99 commented:
https://www.loom.com/share/2982e3d1f3fb4a959e46ab9987e1ef94

I have used the 6digit component.

One doubt, do we need the verifyCode.handler.ts (packages\trpc\server\routers\viewer\organizations\verifyCode.handler.ts) to be a protected route?

@PeerRich @sean-brydon @Udit-takkar

Commit:
prisma schema styling

rkreddy99 commented:
Hey @Udit-takkar , is there anything else to be implemented in the flow?

Udit-takkar commented on packages/features/ee/organizations/components/CreateANewOrganizationForm.tsx:
```diff
@@ -119,10 +138,15 @@ export const VerifyCodeDialog = ({
-                      code: value,
-                      email,
-                    });
+                    isUserSessionRequiredToVerify
```
we require only logged in users to create organisation so i'll undo the changes in this file.

Commit:
refactor: code

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: book event form

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: type error

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: type errors

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Commit:
fix: unit tests

Signed-off-by: Udit Takkar <udit.07814802719@cse.mait.ac.in>

Udit-takkar reviewed (dismissed):
@rkreddy99 LGTM Awesome work.  I pushed some changes

- moved verifyCode dialog to UI
- refactored BookEventForm
- Fixed Type errors
- Added rate limiter in verifycode routes
- some missing translations

sean-brydon commented on packages/ui/components/dialog/VerifyCodeDialog.tsx:
NIT: This should be in features. UI should be dumb components only. This is doing org only logic

sean-brydon reviewed (changes requested):
Over all LGTM - lets keep our UI package pretty dumb and not make calls to trpc where we can

Nice work both of you tho <3 @Udit-takkar @rkreddy99

Commit:
refactor: move verifycodedialog from ui and to features/bookings

rkreddy99 commented:
moved the new verifycodedialog from ui to booking/components

rkreddy99 commented:
> lets keep our UI package pretty dumb 

moved verifycodedialog from calccom/ui to features/booking/components

> not make calls to trpc where we can

added 2 trpc calls
 - sending code for verification thru email
 - verifying the code, unprotected path.

sean-brydon reviewed (dismissed):
LGTM! - Approving - awesome work bud <3 

- I will fix the type error :)

Commit:
fix: type error

rkreddy99 committed:
Merge branch 'main' into rkreddy/9901

rkreddy99 commented:
> * I will fix the type error :)

missed this. fixed it.

sean-brydon reviewed (approved):
Thanks for fixing :)

emrysal committed:
Merge branch 'main' into rkreddy/9901

baileypumfleet commented:
/tip $100

rkreddy99 commented:
I dont know how to claim this, trying out few commands 😅, pls ignore the comment

/claim #9901 
/claim $100
/claim #10317
